dali/base/dadfs.cpp

```diff
@@ -9551,14 +9545,35 @@ bool removeClusterSpares(const char *clusterName, const char *type, SocketEndpoi
     return init.removeSpares(clusterName, type, eps, response);
 }
 
-IGroup *getClusterGroup(const char *clusterName, const char *type, bool expand, unsigned timems)
+IGroup *getClusterNodeGroup(const char *clusterName, const char *type, unsigned timems)
 {
-    CInitGroups init(timems);
-    VStringBuffer cluster("/Environment/Software/%s[@name=\"%s\"]", type, clusterName);
-    Owned<IRemoteConnection> conn = querySDS().connect(cluster.str(), myProcessSession(), RTM_LOCK_READ, SDS_CONNECT_TIMEOUT);
+    VStringBuffer clusterPath("/Environment/Software/%s[@name=\"%s\"]", type, clusterName);
+    Owned<IRemoteConnection> conn = querySDS().connect(clusterPath.str(), myProcessSession(), RTM_LOCK_READ, SDS_CONNECT_TIMEOUT);
     if (!conn)
         return NULL;
-    return init.getGroupFromCluster(type, *conn->queryRoot(), expand);
+    IPropertyTree &cluster = *conn->queryRoot();
+    StringBuffer nodeGroupName;
+    getClusterGroupName(cluster, nodeGroupName);
+    if (0 == nodeGroupName.length())
+        throwUnexpected();
+
+    /* NB: Due to the way node groups and swapNode work, we need to return the IP's from the node group corresponding to the cluster
+     * which may no longer match the cluster IP's due to node swapping.
+     * As the node group is an expanded form of the cluster group (with a IP per partition/slave), with the cluster group repeated
+     * N times, where N is slavesPerNode*channelsPerSlave, return the first M (cluster group width) IP's of the node group.
+     * Ideally the node group representation would change to match the cluster group definition, but that require a lot of changes
+     * to DFS and elsewhere.
+     */
+    Owned<IGroup> nodeGroup = queryNamedGroupStore().lookup(nodeGroupName);
+    CInitGroups init(timems);
+    Owned<IGroup> expandedClusterGroup = init.getGroupFromCluster(type, cluster, true);
+    if (nodeGroup->ordinality() != expandedClusterGroup->ordinality()) // sanity check
+        throwUnexpected();
```

Comment:
minor: Slightly better to use throwUnexpectedX() to give a reason for the failure.